pkg/io/broadcast.go

```diff
@@ -0,0 +1,136 @@
+)
+
+var errEmptySource = fmt.Errorf("Source can't be nil")
+
+type broadcasterData struct {
+	data  interface{}
+	count uint32
+	err   error
+}
+
+type broadcasterRing struct {
+	buffer []atomic.Value
+	// reading (1 bit) + reserved (31 bits) + data count (32 bits)
+	state uint64
+}
+
+func newBroadcasterRing() *broadcasterRing {
+	return &broadcasterRing{buffer: make([]atomic.Value, broadcasterRingSize)}
+}
+
+func (ring *broadcasterRing) index(count uint32) int {
+	return int(count) % len(ring.buffer)
+}
+
+func (ring *broadcasterRing) acquire(count uint32) func(*broadcasterData) {
+	// Reader has reached the latest data, should read from the source.
+	// Only allow 1 reader to read from the source. When there are more than 1 readers,
+	// the other readers will need to share the same data that the first reader gets from
+	// the source.
+	state := uint64(count)
+	if atomic.CompareAndSwapUint64(&ring.state, state, state|maskReading) {
+		return func(data *broadcasterData) {
+			i := ring.index(count)
+			ring.buffer[i].Store(data)
+			atomic.StoreUint64(&ring.state, uint64(count+1))
+		}
+	}
+
+	return nil
+}
+
+func (ring *broadcasterRing) get(count uint32) *broadcasterData {
+	for {
+		reading := uint64(count) | maskReading
+		// TODO: since it's lockless, it spends a lot of resources in the scheduling.
+		for atomic.LoadUint64(&ring.state) == reading {
+			// Yield current goroutine to let other goroutines to run instead
+			time.Sleep(broadcasterRingPollDuration)
+		}
+
+		i := ring.index(count)
+		data := ring.buffer[i].Load().(*broadcasterData)
+		if data.count == count {
+			return data
+		}
+
+		count++
+	}
+}
+
+func (ring *broadcasterRing) lastCount() uint32 {
+	// ring.state always keeps track the next count, so we need to subtract it by 1 to get the
+	// last count
+	return uint32(atomic.LoadUint64(&ring.state)) - 1
+}
+
+// Broadcaster is a generic pull-based broadcaster. Broadcaster is unique in a sense that
+// readers can come and go at anytime, and readers don't need to close or notify broadcaster.
+type Broadcaster struct {
+	source atomic.Value
+	buffer *broadcasterRing
+}
+
+// NewNewBroadcaster creates a new broadcaster.
+func NewBroadcaster(source Reader) *Broadcaster {
+	var broadcaster Broadcaster
+	broadcaster.buffer = newBroadcasterRing()
+	broadcaster.ReplaceSource(source)
+
+	return &broadcaster
+}
+
+// NewReader creates a new reader. Each reader will retrieve the same data from the source.
+// copyFn is used to copy the data from the source to individual readers. Broadcaster uses a small ring
+// buffer, this means that slow readers might miss some data if they're really late and the data is no longer
+// in the ring buffer.
+func (broadcaster *Broadcaster) NewReader(copyFn func(interface{}) interface{}) Reader {
+	currentCount := broadcaster.buffer.lastCount()
+
+	return ReaderFunc(func() (data interface{}, err error) {
+		currentCount++
+		if push := broadcaster.buffer.acquire(currentCount); push != nil {
```

Comment:
If there is a delay on consumer side, provider seems be blocked until consumer is unblocked. (Seen in the metrics of the test on my branch)
Is it intended behavior?

Reply:
Yes, it is intended. The main reasoning is to not waste CPU cycles when there's no reader (on demand). Do you have a concern about blocking the provider?

Reply:
I think once the consumer is delayed during receiving for some reasons, all stream after it will be delayed and never catch up.

Reply:
Oh weird... Are you seeing that from the unit tests? I tried with my mac and ubuntu. They worked fine.

Reply:
I just think it may happens if Broadcaster source reader doesn't check time and drop when Read call is delayed.
All source reader should do something like https://github.com/pion/mediadevices/commit/186ee091027b9550a445d0cc7a641baec3ddea76 to avoid it.

Reply:
Oh right. I think as the source, I would expect to have a slow reader. So, it seems to me that this functionality has been implemented in platform specific, e.g. avdoundation, v4l2.

Let's say we don't let the source to decide this, and we'll have a goroutine that keeps reading the source to keep it hot.

Pros:
* Source is always hot
* Simpler implementation
* The control is on the reader side

Cons:
* Each broadcaster will have at least 1 goroutine to keep it active
* Needs to close the broadcaster to cleanup the goroutine
* When there's no reader, it's still going to be busy

I think both approaches will have their pros and cons. I think blocking the source will make the API more elegant and more efficient since there's no extra goroutine.